src/Utils.php

```diff
@@ -64,4 +62,24 @@ public static function idnUriConvert(UriInterface $uri, $options = 0)
 
         return $uri;
     }
+
+    /**
+     * @param string $domain
+     * @param int    $options
+     * @param array  $info
+     *
+     * @return string|false
+     */
+    private static function idnToAsci($domain, $options, &$info = [])
+    {
+        if (\preg_match('%^[ -~]+$%', $domain) === 1) {
+            return $domain;
+        }
+
+        if (\extension_loaded('intl') && defined('INTL_IDNA_VARIANT_UTS46')) {
```

Comment:
you should remove the `extension_loaded` check. That way, if you have only the polyfill installed, you will still call `idn_to_ascii` (from the polyfill) rather than the internal API.

Reply:
There will always be a case where using the polyfill directly will be needed, so anyway...

BUT it's very possible that ppl forced the polyfill to be NOT installed, by using a "replace" for "intl" in their composer.json.

This "if" should add a `|| !class_exists(Idn::class)` to safeguard against this.

Reply:
Hmm, if anyone has done this, we'd expect a fatal error. I think that would be fine, since it indicates the class doesn't exist, and anyone who has forced non-installation will know that that means.

Reply:
require ext-intl + replace symfony/polyfill-intl is going to be very common.
We recommend it when possible for other polyfilled extensions.
A fatal error would be totally WTF I think...

Reply:
Another option would be to copy the polyfill source code into guzzle I guess, and not require the package.

Reply:
Unless symfony would be interested in creating a reuseable package both the polyfill and guzzle could depend on that would be at risk of being replaced at composer install.

Reply:
To support something that dates back from 2011? I would think Guzzle maintainers had better not do this, it would be pure legacy.

I would suggest to not use the polyfill directly.

That would fix everything for people that don't run outdated software. And for those running outdated software, variant 2003 will be used. That's fine. They can upgrade if they have an issue.

Reply:
> interested in creating a reuseable package both the polyfill and guzzle could depend on

we're not :

Reply:
> There will always be a case where using the polyfill directly will be needed, so anyway...

but that case would be the pure legacy one (having ext-intl installed with a legacy ICU version). Currently, polyfill internals are also used for the non-legacy case of not having ext-intl.

And for that legacy case, I would also vote for using the variant 2003 rather than using the polyfill internals.

Reply:
> To support something that dates back to 2011?

Not actually 2011. Some distros from 2019 ship with old versions of the icu library that don't have this constant. This is one of the points of this PR and the older PR this fixes.

Reply:
> > To support something that dates back to 2011?
> 
> Not actually 2011. Some distros from 2019 ship with old versions of the icu library that don't have this constant. This is one of the points of this PR and the older PR this fixes.

For example CentOS 6, which is still supported until 2020-11-30, has ICU 4.2.1.

Reply:
But do recent PHP versions trigger a deprecation for the variant 2003 even when it was compiled without support for the uts46 variant ? that looks weird to me.